tools/cli/src/mentat_cli/input.rs

```diff
@@ -71,6 +78,44 @@ impl InputReader {
         self.reader.is_some()
     }
 
+    pub fn read_file(&mut self, filename: &String) -> Result<Vec<InputResult>, cli::Error> {
+        let path = Path::new(&filename);
+        let display = path.display();
+
+        let f = match File::open(&path) {
+            Err(err) => bail!(cli::ErrorKind::FileError(display.to_string(), err.to_string())),
+            Ok(file) => file,
+        };
+
+        let mut buffer = String::new();
+
+        let file = BufReader::new(&f);
+        let mut cmd_err: Option<edn::ParseError> = None;
+
+        let mut results: Vec<InputResult> = Vec::new();
+        for line in file.lines() {
+            let l = line.unwrap();
+            println!("{}", l);
+            buffer.push_str(&l);
+            let cmd = Command::Transact(buffer.to_string());
+            let (is_complete, err) = cmd.is_complete();
```

Comment:
Further to the `bail!` below, and to the comments I made on the previous review: if you had `is_complete -> Result<bool, SomeErrorType>` then you can just say `is_complete()?` and the early return will be taken care of.  (You might need to set up `error-chain` to define the lifts from parser error to cli error, but that's what it's there for.)  In general, manually managing and propagating errors is a sign things are not converging to an idiomatic expression.